File: 03월 11일 - 스택, 큐, 덱/과제/2504.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <map>

using namespace std;

int calc(string s) {
    int ans = 0;
    stack<char> st;
    map<char, int> num; //괄호 값
    map<char, char> bracket; //괄호 쌍

    bracket[')'] = '(';
    bracket[']'] = '[';
    num['('] = 2;
    num['['] = 3;

    int temp = 1;
    for (int i = 0; i < s.length(); i++) {
        switch (s[i]) {
            case '(':
            case '[':
                st.push(s[i]);
                temp *= num[s[i]]; //여는 괄호 시 곱하기
                break;
            case ')':
            case ']':
                if (st.empty() || st.top() != bracket[s[i]]) { //올바르지 않은 괄호
                    return 0;
                }
                if (s[i - 1] == bracket[s[i]]) {
                    ans += temp;
                }
                temp /= num[bracket[s[i]]];
                st.pop();
                break;
        }
    }
    if (st.empty()) {
        return ans;
    }
    return 0;
}

int main() {
    string s;

    //입력
    cin >> s;
    //연산 & 출력
    cout << calc(s);
    return 0;
} 

File: 03월 11일 - 스택, 큐, 덱/과제/2504.py
import sys
input = sys.stdin.readline

def calc(text):
    bracket = dict()    # 괄호 쌍 저장
    bracket[')'] = '('
    bracket[']'] = '['
    value = dict()      # 괄호 값 저장
    value['('] = 2
    value['['] = 3

    stack = []
    answer = 0
    weight = 1  # 누적값 -> 밖을 감싸고 있는 괄호들의 곱
    prev = None     # 직전 괄호

    for i in text:
        if i == '(' or i =='[':
            # 여는 괄호 시 곱하기
            weight *= value[i]
            stack.append(i)
        else:
            if len(stack) == 0 or stack[-1] != bracket[i]:  # 올바르지 않은 괄호
                return 0
            if prev == bracket[i]:      # (), [] 형태가 있다면, 현재까지 누적된 값을 더해줌
                answer += weight
            
            # 괄호가 닫혔으므로, 누적값 나누기
            weight //= value[bracket[i]]
            stack.pop()
        prev = i

    if len(stack) == 0:
        return answer
    
    return 0

s = input().rstrip()
print(calc(s))